rmw_connext_dynamic_cpp/src/functions.cpp

```diff
@@ -995,39 +1163,88 @@ void _take(DDS_DynamicData * dynamic_data, void * ros_message,
       case rosidl_typesupport_introspection_cpp::ROS_TYPE_MESSAGE:
         {
           DDS_DynamicData sub_dynamic_data(0, DDS_DYNAMIC_DATA_PROPERTY_DEFAULT);
-          dynamic_data->bind_complex_member(
+          DDS_ReturnCode_t status = dynamic_data->bind_complex_member(
             sub_dynamic_data,
             0,
             i + 1);
+          if (status != DDS_RETCODE_OK) {
+            rmw_set_error_string("failed to bind complex member");
+            return false;
+          }
           void * sub_ros_message = (void *)((char *)ros_message + member->offset_);
           const::rosidl_typesupport_introspection_cpp::MessageMembers * sub_members =
             (const::rosidl_typesupport_introspection_cpp::MessageMembers *)member->members_->data;
-          _take(&sub_dynamic_data, sub_ros_message, sub_members);
-          dynamic_data->unbind_complex_member(sub_dynamic_data);
+          if (!sub_members) {
+            rmw_set_error_string("sub members handle is null");
+            return false;
+          }
+          bool success = _take(&sub_dynamic_data, sub_ros_message, sub_members);
+          status = dynamic_data->unbind_complex_member(sub_dynamic_data);
+          if (!success) {
+            rmw_set_error_string("failed to take sub message");
+            return false;
+          }
+          if (status != DDS_RETCODE_OK) {
+            rmw_set_error_string("failed to unbind complex member");
+            return false;
+          } \
         }
         break;
       default:
-        printf("unknown type id %u\n", member->type_id_);
-        throw std::runtime_error("unknown type");
+        throw std::runtime_error(
+          std::string("unknown type id ") + std::to_string(member->type_id_));
```

Comment:
Should this be an exception? Is it caught before passing through the `rmw` interface?